src/OrleansRuntime/Core/GrainMethodInvoker.cs

```diff
@@ -0,0 +1,183 @@
+        private readonly InvokeInterceptor deprecatedInvokeInterceptor;
+        private readonly InterfaceToImplementationMappingCache interfaceToImplementationMapping;
+        private int stage;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="GrainMethodInvoker"/> class.
+        /// </summary>
+        /// <param name="grain">The grain.</param>
+        /// <param name="request">The request.</param>
+        /// <param name="rootInvoker">The generated invoker.</param>
+        /// <param name="interceptors">The invocation interceptors.</param>
+        /// <param name="interfaceToImplementationMapping">The implementation map.</param>
+        /// <param name="invokeInterceptor">The deprecated silo-wide interceptor.</param>
+        public GrainMethodInvoker(
+            IAddressable grain,
+            InvokeMethodRequest request,
+            IGrainMethodInvoker rootInvoker,
+            List<IMethodInvocationInterceptor> interceptors,
+            InterfaceToImplementationMappingCache interfaceToImplementationMapping,
+            InvokeInterceptor invokeInterceptor)
+        {
+            this.request = request;
+            this.rootInvoker = rootInvoker;
+            this.Grain = grain;
+            this.interceptors = interceptors;
+            this.deprecatedInvokeInterceptor = invokeInterceptor;
+            this.interfaceToImplementationMapping = interfaceToImplementationMapping;
+        }
+
+        /// <inheritdoc />
+        public IAddressable Grain { get; }
+
+        /// <inheritdoc />
+        public MethodInfo Method
+        {
+            get
+            {
+                // Determine if the object being invoked is a grain or a grain extension.
+                Type implementationType;
+                var extensionMap = this.rootInvoker as IGrainExtensionMap;
+                IGrainExtension extension;
+                if (extensionMap != null && extensionMap.TryGetExtension(request.InterfaceId, out extension))
+                {
+                    implementationType = extension.GetType();
+                }
+                else
+                {
+                    implementationType = this.Grain.GetType();
+                }
+
+                // Get or create the implementation map for this object.
+                var implementationMap = interfaceToImplementationMapping.GetOrCreate(
+                    implementationType,
+                    request.InterfaceId);
+
+                // Get the method info for the method being invoked.
+                MethodInfo methodInfo;
+                implementationMap.TryGetValue(request.MethodId, out methodInfo);
+                return methodInfo;
+            }
+        }
+
+        /// <inheritdoc />
+        public object[] Arguments => request.Arguments;
+        
+        /// <inheritdoc />
+        public object Result { get; set; }
+
+        /// <inheritdoc />
+        public async Task Invoke()
+        {
```

Comment:
If any of the interceptor sets the Result property, it should short-circuit the state machine, right?